src/self-update/README.md

```diff
@@ -0,0 +1,66 @@
+# Introduction
+
+That document specifies the feature self-update for thin-edge.
+
+The self-update allows a Cloud Operator to update one thin-edge version that is installed on a device to a newer version, via the Cloud connection. The feature is referred as self-update cause that update process is managed by thin-edge own's Software Management.
+
+# Requirements
+
+* Be default self-update shall use Debian Package Manager APT.
+  * NOTE: APT Package Manager provides a very basic rollback mechanism that still can fail (e.g. due to some unexpected device state or power-loss).
+* Device Owner shall be able to use a power-fail safe filesystem and transaction-safe package manager (instead of APT) - if the device support any.
+  * Therefore an extensible plugin shall be provided.
+* In thin-edge package's installation logic no external tools (e.g. grep, sed, ...) shall be used. Instead Rust code/libraries shall be used.
+  * This is to reduce external dependencies and so reduce risc for failures due missing tools or wrong version of tools.
+
+# Expected Pre-Conditions 
+
+The device and it's thereon installed thin-edge has to follow precondition below to allow successful self-update:
+
+1) The current (old) thin-edge on the device must be installed with usual thin-edge installation, or at leat as usual thin-edge installation would do. That is to assure new thin-edge version that is about to be installed has all external and internal files, folders and resources (e.g. config files) available.
+2) Deployment of thin-edge and mosquitto must not be changed manually after initial installation (e.g. moving/renaming installed files, moving/renaming configuration files used by thin-edge, renaming users/groups created during installation, ...).
+
+3) Package Dependencies need to be solved using SW Management before self-update. This is since self-update will disable auto-dep solving to reduce additional risk during self-update.
+
+*Hint: Before executing self-update on devices in the field it shall be tested on a kind of golden device that equals the device(s) in the field in system-environment, configuration and thin-edge version.*
+
+
+# Execution Flow and involved Components
+![Sequence Diagram Update SW-list](images/self-update.drawio.svg)
+
+
+# Design Principles
+
+### Run-Time Behaviour
+* Self-Update is provided based on usual thin-egde's Software Management.
+  * A plugin for module type "tedge-update" specifically handles thin-edge self-update.
+* The plugin uses Debian Package Manager APT to upgrade requested thin-edge packages.
+* When executing APT the plugin will disable auto dependency solving.
+  * This is to avoid installing more than the requested packages to reduce risc for potential failures.
+* The plugin shall take all requested packages at once from SM Agent (i.E. Plugin API command "update-list" shall be used).
```

Comment:
I see here a risk of inconsistency, the user being able to update only parts of thin-edge.

Did you consider the option to present thin-edge as a package that can only be updated as a whole?